### PR TITLE
Nn 1626 bulk appointment form v2

### PR DESCRIPTION
**Pull request by JamieDraperUK**

### CODE REVIEW

**Comment by JamieDraperUK** [comment]
Don't merge, please.

File: backend/controllers/bulk-appointments-service.js [hunk 1]
```diff
@@ -21,8 +21,13 @@ const BulkAppointmentsServiceFactory = elite2Api => {
     }
   }
 
+  const addBulkAppointments = async (context, appointmnets) => {
+    await elite2Api.addBulkAppointments(context, appointmnets)
+  }
```
Comment (bell-pa) [inline backend/controllers/bulk-appointments-service.js]: typo ^^^ appointmnets (appointments)

File: backend/controllers/controller.js [hunk 2]
```diff
@@ -95,37 +96,57 @@ const factory = ({
       const mb200 = 200000000
```
Comment (bell-pa) [inline backend/controllers/controller.js]: Does this code belong in the controller or a service?

File: backend/controllers/controller.js [hunk 3]
```diff
@@ -95,37 +96,57 @@ const factory = ({
...
       if (size > mb200) {
+        logger.error(`A file exceeding ${mb200} was rejected, filename ${offenders.originalFilename} size ${size}`)
         res.status(400)
-        res.send('The csv is to lage. Maximum file size is 200MB')
+        res.send('The csv is too large. Maximum file size is 200MB')
         return
       }
 
       const bytes = fs.readFileSync(offenders.path)
       if (isBinaryFileSync(bytes, size)) {
```
Comment (bell-pa) [inline backend/controllers/controller.js]: Should these 'sync' operations be asynchronous? 
The node docs for 'fs' say: In busy processes, the programmer is strongly encouraged to use the asynchronous versions of these calls. The synchronous versions will block the entire process until they complete — halting all connections.

Comment (JamieDraperUK) [reply backend/controllers/controller.js]: Agreed.

File: prisonstaffhub-specs/src/test/groovy/uk/gov/justice/digital/hmpps/prisonstaffhub/specs/AddBulkAppointmentsSpecification.groovy [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                        "agencyId", "LEI",
+
+                )])
+
+        elite2api.stubAppointmentTypes([Map.of("code", "ACTI", "description", "Activities")])
+        elite2api.stubBookingOffenders([Map.of("bookingId", 1, "offenderNo", "A12345", "firstName", "John", "lastName", " Doe")])
+        elite2api.stubPostBulkAppointments()
+
+        given: "I navigate to the add bulk appointments screen"
+        go AddBulkAppointmentsPage.url
```
Comment (bell-pa) [inline prisonstaffhub-specs/src/test/groovy/uk/gov/justice/digital/hmpps/prisonstaffhub/specs/AddBulkAppointmentsSpecification.groovy]: I think this should be:
to AddBulkAppointmentsPage

File: prisonstaffhub-specs/src/test/resources/offenders-for-appointments.csv [hunk 5]
```diff
@@ -0,0 +1,3 @@
+offenderNo,start time (example: 22:00 - 22:05 - 22:10 - 22:15 - 22:20 - 22:25 - 22:30 - 22:35 - 22:40 - 22:45 - 22:50 - 22:55)
+A12345,14:20
```
Comment (bell-pa) [inline prisonstaffhub-specs/src/test/resources/offenders-for-appointments.csv]: Does this mean that the following are valid?

A12345, 22:00 - 22:05 
B12345,22:00-22:05

Why ignore the CSV format?
eg
offenderNo, startTime,endTime
A12345,22:00
A12345,22:00,22:05

File: src/BulkAppointments/BulkAppointmentsContainer.js [hunk 6]
```diff
@@ -43,69 +39,96 @@ class BulkAppointmentsContainer extends Component {
...
-      <Page title="Add bulk appointments">
-        <div>{steps[currentStep]}</div>
+      <Page title="Add bulk appointments" alwaysRender>
+        {complete && (
+          <div>
+            <H3> Appointments have been successfully created. </H3>
+            <div>
+              <Button
+                onClick={() =>
+                  this.setState({
```
Comment (bell-pa) [inline src/BulkAppointments/BulkAppointmentsContainer.js]: I think its good practice to keep these function declarations outside the render() method - so they only get created once.

File: src/BulkAppointments/BulkAppointmentsContainer.js [hunk 7]
```diff
@@ -43,69 +39,96 @@ class BulkAppointmentsContainer extends Component {
...
+                render={({ data, error }) => (
+                  <AppointmentDetailsForm
+                    initialValues={appointment}
+                    now={moment()}
+                    appointmentTypes={data.appointmentTypes}
+                    error={error}
+                    locationTypes={data.locationTypes}
+                    onSuccess={values =>
+                      this.setState(state => ({
+                        ...state,
```
Comment (bell-pa) [inline src/BulkAppointments/BulkAppointmentsContainer.js]: move outside render() ?

File: src/BulkAppointments/BulkAppointmentsContainer.js [hunk 8]
```diff
@@ -43,69 +39,96 @@ class BulkAppointmentsContainer extends Component {
...
+            )}
+            {currentStep === 1 && (
+              <Container>
+                <BackLink onClick={() => this.previousStep()}> Back</BackLink>
+                <ChooseOffenders
+                  onError={error => handleError(error)}
+                  onSuccess={data => {
+                    resetErrorDispatch()
+                    this.setState(state => ({
+                      ...state,
```
Comment (bell-pa) [inline src/BulkAppointments/BulkAppointmentsContainer.js]: move outside render() ?

File: src/BulkAppointments/BulkAppointmentsContainer.js [hunk 9]
```diff
@@ -43,69 +39,96 @@ class BulkAppointmentsContainer extends Component {
...
+                      }))
+                    }
+                  />
+                )}
+              />
+            )}
+            {currentStep === 1 && (
+              <Container>
+                <BackLink onClick={() => this.previousStep()}> Back</BackLink>
+                <ChooseOffenders
```
Comment (bell-pa) [inline src/BulkAppointments/BulkAppointmentsContainer.js]: onClick={this.previousStep} ?

File: src/BulkAppointments/BulkAppointmentsContainer.js [hunk 10]
```diff
@@ -43,69 +39,96 @@ class BulkAppointmentsContainer extends Component {
...
+                  onError={error => handleError(error)}
+                  onSuccess={data => {
```
Comment (bell-pa) [inline src/BulkAppointments/BulkAppointmentsContainer.js]: onError={handleError} ?

**Pull request merged**